src/client/Client.cc

```diff
@@ -9243,14 +9244,16 @@ void Client::seekdir(dir_result_t *dirp, loff_t offset)
 //};
 void Client::fill_dirent(struct dirent *de, const char *name, int type, uint64_t ino, loff_t next_off)
 {
-  strncpy(de->d_name, name, 255);
-  de->d_name[255] = '\0';
+  size_t len = strlen(name);
+  len = std::min(len, (size_t)255);
+  memcpy(de->d_name, name, len);
+  de->d_name[len] = '\0';
 #if !defined(__CYGWIN__) && !(defined(_WIN32))
   de->d_ino = ino;
 #if !defined(__APPLE__) && !defined(__FreeBSD__)
   de->d_off = next_off;
 #endif
-  de->d_reclen = 1;
+  de->d_reclen = (uintptr_t)&de->d_name[len] - (uintptr_t)de + 1;
```

Comment:
From my understanding of the man page, this should be:

```suggestion
  de->d_reclen = std::max(sizeof *de, (uintptr_t)&de->d_name[len] - (uintptr_t)de + 1);
```
no? If `de->d_name` is a `char[]` with fixed size, then the `d_reclen == sizeof *de`?

Reply:
Not really. Actually `d_name` field of `struct dirent` is defined as `char d_name[256]` (at least in linux), so using the maximum of `sizeof(*de)` and anything else will return a bigger than necessary record length most of the times.

I've checked the details that man page says about `d_name` and I've found this:

```
The d_name field
       The dirent structure definition shown above is taken from the glibc headers, and shows the d_name field with a fixed size.

       Warning: applications should avoid any dependence on the size of the d_name field.  POSIX defines it as char d_name[], a character array of unspecified  size,
       with at most NAME_MAX characters preceding the terminating null byte ('\0').

       POSIX.1  explicitly  notes  that  this  field  should  not  be  used  as  an lvalue.  The standard also notes that the use of sizeof(d_name) is incorrect; use
       strlen(d_name) instead.  (On some systems, this field is defined as char d_name[1]!)  By implication, the use sizeof(struct dirent) to capture the size of the
       record including the size of d_name is also incorrect.

       Note that while the call

           fpathconf(fd, _PC_NAME_MAX)

       returns the value 255 for most filesystems, on some filesystems (e.g., CIFS, Windows SMB servers), the null-terminated filename that is  (correctly)  returned
       in  d_name  can  actually  exceed this size.  In such cases, the d_reclen field will contain a value that exceeds the size of the glibc dirent structure shown
       above.
```

So it seems clear that using `sizeof` is not correct, but it also says that names could be longer than 255 bytes. Does CephFS enforce a maximum file name of 255 bytes during creations ? otherwise we might need to return longer names here.

Reply:
> Not really. Actually `d_name` field of `struct dirent` is defined as `char d_name[256]` (at least in linux), so using the maximum of `sizeof(*de)` and anything else will return a bigger than necessary record length most of the times.
> 
> I've checked the details that man page says about `d_name` and I've found this:
> 
> ```
> The d_name field
>        The dirent structure definition shown above is taken from the glibc headers, and shows the d_name field with a fixed size.
> 
>        Warning: applications should avoid any dependence on the size of the d_name field.  POSIX defines it as char d_name[], a character array of unspecified  size,
>        with at most NAME_MAX characters preceding the terminating null byte ('\0').
> 
>        POSIX.1  explicitly  notes  that  this  field  should  not  be  used  as  an lvalue.  The standard also notes that the use of sizeof(d_name) is incorrect; use
>        strlen(d_name) instead.  (On some systems, this field is defined as char d_name[1]!)  By implication, the use sizeof(struct dirent) to capture the size of the
>        record including the size of d_name is also incorrect.
> 
>        Note that while the call
> 
>            fpathconf(fd, _PC_NAME_MAX)
> 
>        returns the value 255 for most filesystems, on some filesystems (e.g., CIFS, Windows SMB servers), the null-terminated filename that is  (correctly)  returned
>        in  d_name  can  actually  exceed this size.  In such cases, the d_reclen field will contain a value that exceeds the size of the glibc dirent structure shown
>        above.
> ```
> 
> So it seems clear that using `sizeof` is not correct, but it also says that names could be longer than 255 bytes.

Right. So we should remember the whole point of `d_reclen` is so that we can efficiently pack dirents returned by readdir via the syscall. This is obviously not an issue for a userspace implementation (where we're not packing dirents). I do wonder however if `d_reclen` should be word-aligned.

>  Does CephFS enforce a maximum file name of 255 bytes during creations ? otherwise we might need to return longer names here.

It does IIRC. It was actually an issue for fscrypt which is why we have alternate_name.

Reply:
> Right. So we should remember the whole point of `d_reclen` is so that we can efficiently pack dirents returned by readdir via the syscall. This is obviously not an issue for a userspace implementation (where we're not packing dirents). I do wonder however if `d_reclen` should be word-aligned.

Even thought we are not packing more than one dirent, the definition of the field says that it contains the length + 1 of the entry name, so I think we should preserve that to be compliant with the specification.

In this case `d_reclen` doesn't need to be aligned to word size, it needs to be exactly the length of the name + 1. In requests that use an array of dirents instead of just one, it's the array itself that aligns each entry to the appropriate size (handled by the compiler).

> > Does CephFS enforce a maximum file name of 255 bytes during creations ? otherwise we might need to return longer names here.
> 
> It does IIRC. It was actually an issue for fscrypt which is why we have alternate_name.

It's not related to this patch, but given that the man page says that `d_reclen` can actually be larger than 255, couldn't we just return the `alternate_name` as the entry name when it's longer than 255 ? this way it would be transparent for the clients.

Reply:
Sounds like we did not reach a conclusion on this @batrick @xhernandez ??

Reply:
> > Right. So we should remember the whole point of `d_reclen` is so that we can efficiently pack dirents returned by readdir via the syscall. This is obviously not an issue for a userspace implementation (where we're not packing dirents). I do wonder however if `d_reclen` should be word-aligned.
> 
> Even thought we are not packing more than one dirent, the definition of the field says that it contains the length + 1 of the entry name, so I think we should preserve that to be compliant with the specification.

The actual requirements come from this standard:

https://pubs.opengroup.org/onlinepubs/9799919799.2024edition/functions/posix_getdents.html

particularly:

> The value of the d_reclen member shall be set to the number of bytes occupied by this entry in buf, including any padding bytes needed before the next entry, if any. If this is the last entry in buf, d_reclen shall include any padding bytes needed to make the address of this entry plus d_reclen bytes satisfy the alignment requirements for the posix_dent structure.

So I think alignment should be considered when setting the length.

> > > Does CephFS enforce a maximum file name of 255 bytes during creations ? otherwise we might need to return longer names here.
> > 
> > 
> > It does IIRC. It was actually an issue for fscrypt which is why we have alternate_name.
> 
> It's not related to this patch, but given that the man page says that `d_reclen` can actually be larger than 255, couldn't we just return the `alternate_name` as the entry name when it's longer than 255 ? this way it would be transparent for the clients.

No. `alternate_name` is not an application visible member of the dentry metadata. It exists to make encryption of large file names possible.

Reply:
Yes. You are right. I incorrectly interpreted that the value was based just on the length of the name. I've updated the patch to address the alignment requirement.